src/lib/plugin.js

```diff
@@ -805,21 +807,46 @@ class FiveBellsLedger extends EventEmitter2 {
   }
 
   * _requestWithCredentials (options) {
-    const authToken = yield this._getAuthToken()
-    return yield request(Object.assign(
-      requestCredentials(this.credentials, authToken),
-      options))
+    // figure out which auth mechanism is supported
+    if (!this.supportedAuth) {
+      this.supportedAuth = yield this._getAuthMechanisms()
+    }
+
+    let requestCreds = null
+    if (this.supportedAuth === 'token') {
+      requestCreds = requestCredentials(this.credentials, yield this._getAuthToken())
+    } else if ((this.supportedAuth === 'basic')) {
+      requestCreds = requestCredentials(this.credentials)
+    }
+
+    return yield request(Object.assign(requestCreds, options))
+  }
+
+  * _getAuthMechanisms () {
+    let tryAuth = function * (options) {
+      let resp = yield request.post(this.ledgerContext.urls.message, options)
+      return +resp.statusCode
+    }.bind(this)
+
+    if ((yield tryAuth({auth: {bearer: 'invalidToken'}})) === 403) {
+      return 'token'
+    } else if ((yield tryAuth({auth: {user: 'abc', pass: 'abc'}})) === 403) {
```

Comment:
Concerning this code snippet, we can ignore other auth methods like client certs. It only has to decide if it should use a bearer token or basic auth (they are mutually exclusive).

If a (future/3rd party) ledger implementation supports neither bearer token nor basic auth, we will get a clear error message stating this fact. Doing this one additional check does not impose any noteworthy overhead (performance or otherwise), so why not keep it for the sake of clear error messages and "failing early".